File: app/src/main/java/com/etesync/syncadapter/journalmanager/BaseManager.java
package com.etesync.syncadapter.journalmanager;

import com.etesync.syncadapter.App;
import com.etesync.syncadapter.GsonHelper;

import org.apache.commons.codec.Charsets;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.net.HttpURLConnection;
import java.util.logging.Level;

import okhttp3.HttpUrl;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;

abstract class BaseManager {
    final static protected MediaType JSON = MediaType.parse("application/json; charset=utf-8");

    protected HttpUrl remote;
    protected OkHttpClient client;

    Response newCall(Request request) throws Exceptions.HttpException {
        Response response;
        try {
            response = client.newCall(request).execute();
        } catch (IOException e) {
            App.log.log(Level.SEVERE, "Failed while connecting to server", e);
            throw new Exceptions.ServiceUnavailableException("[" + e.getClass().getName() + "] " + e.getLocalizedMessage());
        }

        if (!response.isSuccessful()) {
            ApiError apiError = GsonHelper.gson.fromJson(response.body().charStream(), ApiError.class);

            switch (response.code()) {
                case HttpURLConnection.HTTP_UNAVAILABLE:
                    throw new Exceptions.ServiceUnavailableException(response, "Service unavailable");
                case HttpURLConnection.HTTP_UNAUTHORIZED:
                    throw new Exceptions.UnauthorizedException(response, "Unauthorized auth token");
                case HttpURLConnection.HTTP_FORBIDDEN:
                    if (apiError.code.equals("service_inactive")) {
                        throw new Exceptions.UserInactiveException(response, apiError.detail);
                    }
                default:
                    // Fall through. We want to always throw when unsuccessful.
            }

            throw new Exceptions.HttpException(response, apiError.detail);
        }

        return response;
    }

    static class ApiError {
        String detail;
        String code;

        ApiError() {
        }
    }

    static class Base {
        private byte[] content;
        private String uid;

        void setContent(final byte[] content) {
            this.content = content;
        }

        byte[] getContent() {
            return this.content;
        }

        void setUid(final String uid) {
            this.uid = uid;
        }

        public String getUid() {
            return uid;
        }

        public String getContent(Crypto.CryptoManager crypto) {
            return new String(crypto.decrypt(content), Charsets.UTF_8);
        }

        void setContent(Crypto.CryptoManager crypto, String content) {
            this.content = crypto.encrypt(content.getBytes(Charsets.UTF_8));
        }

        byte[] calculateHmac(Crypto.CryptoManager crypto, String uuid) {
            ByteArrayOutputStream hashContent = new ByteArrayOutputStream();

            try {
                if (uuid != null) {
                    hashContent.write(uuid.getBytes(Charsets.UTF_8));
                }

                hashContent.write(content);
            } catch (IOException e) {
                // Can never happen, but just in case, return a bad hmac
                return "DEADBEEFDEADBEEFDEADBEEFDEADBEEF".getBytes();
            }

            return crypto.hmac(hashContent.toByteArray());
        }

        protected Base() {
        }

        Base(Crypto.CryptoManager crypto, String content, String uid) {
            setContent(crypto, content);
            setUid(uid);
        }

        @Override
        public String toString() {
            return getClass().getSimpleName() + "<" + uid + ">";
        }

        String toJson() {
            return GsonHelper.gson.toJson(this, getClass());
        }
    }
}

File: app/src/main/java/com/etesync/syncadapter/journalmanager/BaseManager.kt
package com.etesync.syncadapter.journalmanager

import com.etesync.syncadapter.GsonHelper
import com.etesync.syncadapter.log.Logger
import okhttp3.*
import java.io.ByteArrayOutputStream
import java.io.IOException
import java.net.HttpURLConnection
import java.util.logging.Level

abstract class BaseManager {

    protected var remote: HttpUrl? = null
    protected var client: OkHttpClient? = null

    @Throws(Exceptions.HttpException::class)
    fun newCall(request: Request): Response {
        val response: Response
        try {
            Logger.log.fine("Making request for ${request.url()}")
            response = client!!.newCall(request).execute()
        } catch (e: IOException) {
            Logger.log.log(Level.SEVERE, "Failed while connecting to server", e)
            throw Exceptions.ServiceUnavailableException("[" + e.javaClass.name + "] " + e.localizedMessage)
        }

        if (!response.isSuccessful) {
            val apiError = if (response.header("Content-Type", "application/json")!!.startsWith("application/json"))
                GsonHelper.gson.fromJson(response.body()!!.charStream(), ApiError::class.java)
            else
                ApiError(code="got_html", detail="Got HTML while expecting JSON")


            when (response.code()) {
                HttpURLConnection.HTTP_BAD_GATEWAY -> throw Exceptions.BadGatewayException(response, "Bad gateway: most likely a server restart")
                HttpURLConnection.HTTP_UNAVAILABLE -> throw Exceptions.ServiceUnavailableException(response, "Service unavailable")
                HttpURLConnection.HTTP_UNAUTHORIZED -> throw Exceptions.UnauthorizedException(response, "Unauthorized auth token")
                HttpURLConnection.HTTP_CONFLICT -> throw Exceptions.ConflictException(response, "Http conflict")
                HttpURLConnection.HTTP_FORBIDDEN -> {
                    if (apiError.code == "service_inactive") {
                        throw Exceptions.UserInactiveException(response, apiError.detail)
                    } else if (apiError.code == "associate_not_allowed") {
                        throw Exceptions.AssociateNotAllowedException(response, apiError.detail)
                    } else if (apiError.code == "journal_owner_inactive") {
                        throw Exceptions.ReadOnlyException(response, apiError.detail)
                    }
                }
            }// Fall through. We want to always throw when unsuccessful.

            throw Exceptions.HttpException(response, apiError.detail)
        }

        return response
    }

    internal class ApiError(
        var detail: String? = null,
        var code: String? = null
    )

    open class Base {
        var content: ByteArray? = null
        var uid: String? = null

        fun getContent(crypto: Crypto.CryptoManager): String {
            return String(crypto.decrypt(content!!)!!)
        }

        fun setContent(crypto: Crypto.CryptoManager, content: String) {
            this.content = crypto.encrypt(content.toByteArray())
        }

        fun calculateHmac(crypto: Crypto.CryptoManager, uuid: String?): ByteArray {
            val hashContent = ByteArrayOutputStream()

            try {
                if (uuid != null) {
                    hashContent.write(uuid.toByteArray())
                }

                hashContent.write(content!!)
            } catch (e: IOException) {
                // Can never happen, but just in case, return a bad hmac
                return "DEADBEEFDEADBEEFDEADBEEFDEADBEEF".toByteArray()
            }

            return crypto.hmac(hashContent.toByteArray())
        }

        protected constructor() {}

        constructor(crypto: Crypto.CryptoManager, content: String, uid: String) {
            setContent(crypto, content)
            this.uid = uid
        }

        override fun toString(): String {
            return javaClass.simpleName + "<" + uid + ">"
        }

        internal open fun toJson(): String {
            return GsonHelper.gson.toJson(this, javaClass)
        }
    }

    companion object {
        val JSON = MediaType.parse("application/json; charset=utf-8")
    }
}
